### PR TITLE
feat: Add `signal` option so that requests can be aborted

### LINKED ISSUE #891: Allow requests to be aborted
`RequestOptions` should include a `signal: AbortSignal`, and requests should be aborted when the signal is aborted.

### PR DESCRIPTION
**Pull request by haines**
Closes #891

### CODE REVIEW

File: private/test/src/client/abort.test.ts [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+      try {
+        await response;
+        throw new Error("expected an error to be thrown");
+      } catch (error) {
+        expect(error).toMatchObject({
+          constructor: NotOK,
+          code: Status.CANCELLED,
+          cause: reason,
+        });
+      }
```
Comment (hasanayan) [inline private/test/src/client/abort.test.ts:54]: How about this one?

```suggestion
      await expect(response).rejects.toMatchObject({
        constructor: NotOK,
        code: Status.CANCELLED,
        cause: reason,
      });
```

**Review by hasanayan — APPROVED** [review]

File: private/test/src/client/abort.test.ts (hunk 1 above)
Comment (hasanayan) [reply private/test/src/client/abort.test.ts:54]: if you accept this, please consider it for the next throws assertions

File: packages/core/src/errors.ts [hunk 2]
```diff
@@ -127,7 +140,11 @@ export class ValidationFailed extends Error {
...
     super("Input failed schema validation");
-    this.name = this.constructor.name;
-    Error.captureStackTrace(this, this.constructor);
+    setNameAndStack(this);
   }
 }
+
+function setNameAndStack(error: Error): void {
+  error.name = error.constructor.name;
+  Error.captureStackTrace?.(error, error.constructor); // eslint-disable-line @typescript-eslint/no-unnecessary-condition -- `Error.captureStackTrace` is not available in all browsers
```
[the remark this answers is not in the document]
Comment (haines) [reply packages/core/src/errors.ts:149]: Huh, you'd think that the compiler ought to be able to realise that `"captureStackTrace" in Error` is also always true given the types.

[the remark this answers is not in the document]
Comment (hasanayan) [reply packages/core/src/errors.ts]: yeah, I had to try it out to be sure 😄

**Pull request merged**